### PR TITLE
refactor(core): Move push message types to a new shared package (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary
This PR
1. sets up a new package for sharing types (and validatable schemas) between the backend and the frontend
2. improve type-safety on all push methods

## Review / Merge checklist

- [x] PR title and summary are descriptive
- [ ] Tests included

### CODE REVIEW

File: packages/@n8n/api-types/src/push/collaboration.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import type { Iso8601DateTimeString } from '../types';
+
+export interface Collaborator {
+	userId: string;
```
Comment (netroy) [inline packages/@n8n/api-types/src/push/collaboration.ts:4]: port over a `User` type into this package, and revert this back to `user: User` to avoid making unrelated changes in this PR.

File: packages/@n8n/api-types/src/push/collaboration.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	lastSeen: Iso8601DateTimeString;
+}
+
+interface CollaboratorsChanged {
```
Comment (netroy) [inline packages/@n8n/api-types/src/push/collaboration.ts:8]: Add JSDoc on every push message explaining when it's used.

File: packages/cli/src/services/orchestration/main/handle-worker-response-message-main.ts [hunk 3]
```diff
@@ -27,7 +29,7 @@ export async function handleWorkerResponseMessageMain(
 		case 'getStatus':
 			Container.get(Push).broadcast('sendWorkerStatusMessage', {
 				workerId: workerResponse.workerId,
-				status: workerResponse.payload,
+				status: workerResponse.payload as WorkerStatus,
```
Comment (tomi) [inline packages/cli/src/services/orchestration/main/handle-worker-response-message-main.ts:32]: Could we get rid of this cast? Is `payload` correctly typed?

File: packages/cli/src/services/redis/redis-service-commands.ts [hunk 4]
```diff
@@ -42,7 +43,7 @@ export type RedisServiceBaseCommand =
 	| {
 			senderId: string;
 			command: 'relay-execution-lifecycle-event';
-			payload: { type: IPushDataType; args: Record<string, unknown>; pushRef: string };
+			payload: { type: PushType; args: Record<string, unknown>; pushRef: string };
```
Comment (tomi) [inline packages/cli/src/services/redis/redis-service-commands.ts:46]: Is this any `PushType` or should it be only a subset of them?

File: packages/cli/src/databases/entities/installed-nodes.ts [hunk 5]
```diff
@@ -10,7 +10,7 @@ export class InstalledNodes {
 	type: string;
 
 	@Column()
-	latestVersion: string;
+	latestVersion: number;
```
Comment (tomi) [inline packages/cli/src/databases/entities/installed-nodes.ts:13]: This change is kinda unrelated to this PR. Could it be extracted into its own?

File: packages/editor-ui/src/Interface.ts [hunk 6]
```diff
@@ -422,213 +420,6 @@ export interface IExecutionDeleteFilter {
...
-		internal: boolean;
-	}>;
-	version: string;
-}
-
-export interface IPushDataWorkerStatusMessage {
-	workerId: string;
-	status: IPushDataWorkerStatusPayload;
-}
-
```
Comment (tomi) [inline packages/editor-ui/src/Interface.ts:631]: image

File: packages/editor-ui/src/components/MainHeader/CollaborationPane.vue [hunk 7]
```diff
@@ -21,13 +20,11 @@ watch(visibility, (visibilityState) => {
...
 
 const collaboratorsSorted = computed(() => {
-	const currentWorkflowUsers = collaborationStore.collaborators.map(({ user }) => user);
-	const owner = currentWorkflowUsers.find(isUserGlobalOwner);
-	return {
-		defaultGroup: owner
-			? [owner, ...currentWorkflowUsers.filter((user) => user.id !== owner.id)]
-			: currentWorkflowUsers,
-	};
+	const users = collaborationStore.collaborators.map(({ user }) => user);
```
Comment (tomi) [inline packages/editor-ui/src/components/MainHeader/CollaborationPane.vue:23]: This is also unrelated to the changes in this PR, could it be extracted?

File: packages/editor-ui/src/composables/usePushConnection.ts [hunk 8]
```diff
@@ -31,19 +28,18 @@ import { useWorkflowsStore } from '@/stores/workflows.store';
...
 import { useExternalHooks } from '@/composables/useExternalHooks';
-import type { useRouter } from 'vue-router';
 import { useWorkflowHelpers } from '@/composables/useWorkflowHelpers';
 import { useI18n } from '@/composables/useI18n';
 import { useTelemetry } from '@/composables/useTelemetry';
 import type { PushMessageQueueItem } from '@/types';
 import { useAssistantStore } from '@/stores/assistant.store';
 import NodeExecutionErrorMessage from '@/components/NodeExecutionErrorMessage.vue';
 
+type IPushDataExecutionFinished = PushPayload<'executionFinished'>;
```
Comment (tomi) [inline packages/editor-ui/src/composables/usePushConnection.ts:41]:
```suggestion
type IPushDataExecutionFinishedPayload = PushPayload<'executionFinished'>;
```
?

**Review by tomi — COMMENTED** [review]
Oh this is really nice, great job! Left a few comments. Also:

- Should we unify the type names to have a `Message` suffix for all them?

File: turbo.json
Comment (ivov) [inline turbo.json:1]: Let's add a small `README.md` to the new package?

**Review by ivov — COMMENTED** [review]
My new fav package 🙏🏻

File: packages/@n8n/api-types/src/types.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+/** DateTime in the Iso8601 format, e.g. 2024-10-31T00:00:00.123Z */
+export type Iso8601DateTimeString = string;
```
[the remark this answers is not in the document]
Comment (tomi) [reply packages/@n8n/api-types/src/types.ts:2]: Interesting idea 👍 Might make it hard to interop with `new Date().toISOString()` tho

File: packages/cli/src/databases/entities/installed-nodes.ts (hunk 5 above)
Comment (netroy) [reply packages/cli/src/databases/entities/installed-nodes.ts:13]: https://github.com/n8n-io/n8n/pull/10782

File: packages/@n8n/api-types/package.json [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  "scripts": {
+    "clean": "rimraf dist .turbo",
+    "dev": "pnpm watch",
+    "typecheck": "tsc --noEmit",
+    "build": "tsc -p tsconfig.build.json",
+    "format": "prettier --write . --ignore-path ../../../.prettierignore",
+    "lint": "eslint . --quiet",
+    "lintfix": "eslint . --fix",
+    "watch": "tsc -p tsconfig.build.json --watch",
+    "test": "echo \"No tests yet\" && exit 0"
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/@n8n/api-types/package.json:13]: in a follow up PR, I'll move some request payloads in this package. and we can add unit tests for all validation scenarios.

File: packages/@n8n/api-types/src/types.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (netroy) [reply packages/@n8n/api-types/src/types.ts]: considering that this always represents an auto-generated value, I don't think it's worth the effort.

File: packages/cli/src/controllers/e2e.controller.ts [hunk 11]
```diff
@@ -131,7 +133,7 @@ export class E2EController {
 	}
 
 	@Post('/push', { skipAuth: true })
-	async pushSend(req: PushRequest) {
+	async pushSend(req: PushRequest<any>) {
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/controllers/e2e.controller.ts:136]: I tried that first, but could not get it to work.

File: packages/cli/src/load-nodes-and-credentials.ts [hunk 12]
```diff
@@ -376,7 +376,7 @@ export class LoadNodesAndCredentials {
 				loader.reset();
 				await loader.loadAll();
 				await this.postProcessLoaders();
-				push.broadcast('nodeDescriptionUpdated');
+				push.broadcast('nodeDescriptionUpdated', {});
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/load-nodes-and-credentials.ts:378]: i don't know how to make the second argument conditional when the first argument is of this type.
perhaps we should change the signature on these functions to accept a single object arg with `type` and `data`, instead of 2 args.

File: packages/cli/src/services/orchestration/main/handle-worker-response-message-main.ts (hunk 3 above)
Comment (netroy) [reply packages/cli/src/services/orchestration/main/handle-worker-response-message-main.ts:33]: I tried, but it turned out to be complicated. I think we need to refactor the redis commands a bit in another PR.

File: packages/cli/src/services/redis/redis-service-commands.ts (hunk 4 above)
Comment (netroy) [reply packages/cli/src/services/redis/redis-service-commands.ts:47]: it should definitely be a subtype, but this also requires a refactor of redis commands code first.

**Review by tomi — DISMISSED** [review]
💘  🚀

**Review by tomi — APPROVED** [review]

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.60.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.60.0)
